Source/SwiftLintFramework/Models/Linter.swift

```diff
@@ -43,14 +44,35 @@ public struct Linter {
                     ruleTimes.append((id, -start.timeIntervalSinceNow))
                 }
             }
+
             return violations.filter { violation in
-                guard let violationRegion = regions
-                    .first(where: { $0.contains(violation.location) }) else {
-                        return true
+                guard let violationRegion = regions.first(where: { $0.contains(violation.location) }) else {
+                    return true
                 }
-                return violationRegion.isRuleEnabled(rule)
+                let enabled = violationRegion.isRuleEnabled(rule)
+                if !enabled {
+                    let identifiers = violationRegion.deprecatedAliasesDisablingRule(rule)
+                    if !identifiers.isEmpty {
+                        mutationQueue.sync {
```

Comment:
I already wasn't very happy about the side-effect mutation in this `flatMap` block (my own fault), but this really brings it too far.

I've sketched out a way to fix this, and reduce locking while we're at it, but it needs some refactoring and better naming:

```diff
diff --git a/Source/SwiftLintFramework/Models/Linter.swift b/Source/SwiftLintFramework/Models/Linter.swift
index d370859..980f7fb 100644
--- a/Source/SwiftLintFramework/Models/Linter.swift
+++ b/Source/SwiftLintFramework/Models/Linter.swift
@@ -27,48 +27,51 @@ public struct Linter {
             queuedPrintError("Most rules will be skipped because sourcekitd has failed.")
         }
         let regions = file.regions()
-        var ruleTimes = [(id: String, time: Double)]()
-        let mutationQueue = DispatchQueue(label: "io.realm.SwiftLintFramework.getStyleViolationsMutation")
-        var deprecatedIdentifiers = Set<String>()
-        var deprecatedToValidIdentifier = [String: String]()
-
-        let violations = rules.parallelFlatMap { rule -> [StyleViolation] in
+        let result = rules.parallelMap { rule -> ([StyleViolation], (String, Double)?, [(String, String)]) in
             if !(rule is SourceKitFreeRule) && self.file.sourcekitdFailed {
-                return []
+                return ([], nil, [])
             }
-            let start: Date! = benchmark ? Date() : nil
-            let violations = rule.validateFile(self.file)
+            let violations: [StyleViolation]
+            let benchmarkValue: (String, Double)?
             if benchmark {
+                let start = Date()
+                violations = rule.validateFile(self.file)
                 let id = type(of: rule).description.identifier
-                mutationQueue.sync {
-                    ruleTimes.append((id, -start.timeIntervalSinceNow))
-                }
+                benchmarkValue = (id, -start.timeIntervalSinceNow)
+            } else {
+                violations = rule.validateFile(self.file)
+                benchmarkValue = nil
             }
 
-            return violations.filter { violation in
-                guard let violationRegion = regions.first(where: { $0.contains(violation.location) }) else {
-                    return true
-                }
-                let enabled = violationRegion.isRuleEnabled(rule)
-                if !enabled {
-                    let identifiers = violationRegion.deprecatedAliasesDisablingRule(rule)
-                    if !identifiers.isEmpty {
-                        mutationQueue.sync {
-                            deprecatedIdentifiers.formUnion(identifiers)
-                            for deprecatedIdentifier in identifiers {
-                                let identifier = type(of: rule).description.identifier
-                                deprecatedToValidIdentifier[deprecatedIdentifier] = identifier
-                            }
-                        }
-                    }
-                }
+            var violationsAndRegions = violations.map { violation in
+                return (violation, regions.first(where: { $0.contains(violation.location) }))
+            }
+
+            let pivot = violationsAndRegions.partition { _, region in
+                return region?.isRuleEnabled(rule) ?? true
+            }
 
-                return enabled
+            let enabledViolations = violationsAndRegions[0..<pivot].map { $0.0 }
+            let disabledViolationsAndRegions = violationsAndRegions[pivot..<violationsAndRegions.count]
+            let thingy = disabledViolationsAndRegions.flatMap { _, region -> [(String, String)] in
+                let identifiers = region?.deprecatedAliasesDisablingRule(rule) ?? []
+                return identifiers.map { ($0, type(of: rule).description.identifier) }
             }
+            return (enabledViolations, benchmarkValue, thingy)
+        }
+
+        let violations = result.flatMap { subResult in
+            return subResult.0
+        }
+        let ruleTimes = result.flatMap { subResult in
+            return subResult.1
+        }
+        var deprecatedToValidIdentifier = [String: String]()
+        for (key, value) in result.flatMap({ $0.2 }) {
+            deprecatedToValidIdentifier[key] = value
         }
 
-        for deprecatedIdentifier in deprecatedIdentifiers {
-            guard let identifier = deprecatedToValidIdentifier[deprecatedIdentifier] else { continue }
+        for (deprecatedIdentifier, identifier) in deprecatedToValidIdentifier {
             queuedPrintError("'\(deprecatedIdentifier)' rule has been renamed to '\(identifier)' and will be " +
                 "completely removed in a future release.")
         }
```